### ISSUE TITLE
[Perf] async_scheduling silently disabled under UniProcExecutor → rollout device-idle (continue_decode coupling?)

### ISSUE THREAD

**Issue by lokic233** [vllm-project/tpu-inference#3456]
### Summary

On the JAX path, vLLM forces `UniProcExecutor` for single-host (no pipeline parallelism). `UniProcExecutor.supports_async_scheduling()` returns `False`, so even when the user requests `async_scheduling: true`, the engine resets it to `False`. As a result the async branch in `TPUModelRunner` (`tpu_inference/runner/tpu_runner.py`, e.g. the `if self.scheduler_config.async_scheduling:` sites around L1840 / L2154) is **dead code** on this configuration, and every rollout wave blocks on `device_get` → large inter-wave host idle.

### Environment

- TPU v7x (Ironwood), 4×4×4, 64 chips, single slice
- Collocated GRPO rollout, per-process-DP `UniProcExecutor` (single host, no PP)
- `async_scheduling: true` requested in config; `enable_continue_decode: true`

### Mechanism

1. Config requests `async_scheduling=true`.
2. `tpu_platform.py` (~L297) forces `UniProcExecutor` for JAX single-host without PP.
3. `UniProcExecutor.supports_async_scheduling()` → `False`, so vLLM resets `scheduler_config.async_scheduling=False`.
4. Therefore `tpu_runner.py`'s `if self.scheduler_config.async_scheduling:` branches never fire → the continue_decode async overlap path is never taken → each wave blocks on `device_get` before enqueuing the next → the rollout phase sits mostly device-idle (we measure ~90% rollout device-idle / dispatch-bound on this shape).

### What we observed (measured, on the #3262 spmd backend)

Forcing `supports_async_scheduling` → `True` (local experiment) revived the async branch and cut the warm rollout chunk by **~15%** (≈15.7s → ≈13.6s per chunk), with correctness preserved on our workload (loss trend, reward distribution, EOS counts, and mean completion length all unchanged vs the sync path under fixed seed).

### The open question (why this is an issue, not a blind PR)

`tpu_runner.py` (~L720) carries an explicit comment that the sync `_sample_from_logits` path is `async_scheduling=False, required by continue_decode` — i.e. the async-off ↔ continue_decode coupling looks **intentional**. Our force-on experiment passed correctness *on our specific workload*, but we don't want to override a documented invariant blindly.

**Question for maintainers:** is `async_scheduling` safe to enable under `UniProcExecutor` when `enable_continue_decode=true`? If the coupling is a hard invariant, is there a narrower way to expose the async inter-wave overlap for the collocated single-host RL rollout case? We're happy to send a PR in whichever direction you prefer (a `supports_async_scheduling` override, or a continue_decode-safe overlap path).

**Comment by lokic233** [comment]
Follow-up after reading the current `main` head (73de2b6) more carefully — the situation is narrower and cleaner than the original report implied, and I'd like maintainer guidance on **which layer** the fix belongs in before I send a PR.

**1. async + continue_decode already coexist in the runner.** There is a live async continue_decode path in `tpu_runner.py` (`_execute_continue_decode` builds an `AsyncTPUModelRunnerOutput` and sets `async_output._is_continue_decode = True`, ~L1912; `_modify_prev_results` handles `is_continue_decode` at ~L1418). So the comment at ~L720 (`async_scheduling=False, required by continue_decode`) reads as a **local invariant of the sync `_sample_from_logits` path**, not a global prohibition on async+continue_decode. Please correct me if that's wrong.

**2. The actual gate is in vLLM core, not tpu-inference.** `tpu_platform.py` forces `distributed_executor_backend="uni"` for JAX single-host w/o PP; vLLM's `UniProcExecutor.supports_async_scheduling()` returns `False`, so vLLM resets `scheduler_config.async_scheduling` `True → False`. tpu-inference only *consumes* the resulting flag. So a tpu-inference-side "force it on" edit would be fighting the layering.

**3. Overlap with #3401.** #3401 ("[continue_decode] Make EOS_CHECK_INTERVAL<=0 a correct continuous-batching window") is reworking the continue_decode host-overhead path for decode-heavy RL post-training — the same host-idle this issue is about. It may already remove much of the inter-wave bubble from a better angle.

**Question:** for the collocated single-host RL rollout case (`UniProcExecutor`, `enable_continue_decode=true`), where would you prefer the async inter-wave overlap to be enabled?
- (a) vLLM core: make `UniProcExecutor.supports_async_scheduling()` return `True` (upstream vLLM PR), or
- (b) tpu-inference platform-side: an opt-in that re-enables `scheduler_config.async_scheduling` after the executor is chosen (default-off env), or
- (c) it's already covered / subsumed by #3401 and no separate change is wanted.

I have a working local change for (a) that gave ~15% off the warm rollout chunk on TPU v7x 4×4×4 (64-chip) collocated GRPO with correctness preserved on that workload (loss/reward/EOS/mean-completion-length parity vs sync under fixed seed) — happy to send it upstream to vLLM, or do (b) here, whichever you prefer. Just don't want to override the documented invariant without a maintainer's read.

**Comment by lokic233** [comment]
**Update — I believe this is already resolved on current `main`, so no code change is needed. Recording the finding in case it helps others hitting the older behavior.**

The gate I described (async silently reset to `False` under `UniProcExecutor`) was against an older vLLM. On the vLLM commit tpu-inference currently pins (`d626108b…`), `UniProcExecutor.supports_async_scheduling()` **already returns `True`**:

```python
# vllm/v1/executor/uniproc_executor.py
@classmethod
def supports_async_scheduling(cls) -> bool:
    return True
```

And `vllm/config/vllm.py` no longer force-resets `async_scheduling → False` for a supported executor — with `async_scheduling: true` it now *keeps* it (raising only for genuinely-incompatible combos like certain spec-decode configs). So on a current stack, requesting `async_scheduling: true` on the single-host `uni` path is honored, and the runner's async inter-wave branch is reachable — which is exactly the ~15% warm-rollout-chunk improvement I measured (via a local override on the older stack).

**Net:** the fix landed upstream in vLLM; tpu-inference just needed the LKG bump that already happened. No tpu-inference code change required for the executor gate itself.

The one thing that may still be worth a doc/example note: for the collocated single-host RL rollout case, `async_scheduling: true` is **not on by default** and is worth calling out as a lever (with the continue_decode interaction validated). If maintainers think a short note in the DAPO/RL example (#3165) is useful, I'm happy to add it there. Otherwise this can be closed as resolved-upstream. Thanks!

**Comment by lokic233** [comment]
Opened #3458 — a self-contained reproduction example (`examples/dapo_tpu_async/`) so the async-overlap result here is citable/reproducible end-to-end on a current vLLM LKG (config + levers + measured ~15% warm-rollout-chunk / ~94s median cycle, correctness-gated). Docs-only; complements #3457 and #3165.
